### ISSUE TITLE
[no-unsafe-assignment] option to ignore object destructuring or shorthand?

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3867]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-assignment": [2]
  }
}
```

```TS
// the following `body` emits the lint error
const { url, query, body } = req  as {url: string, query: string, body: any}

const testFn = <Q, T>({ query, body }: { query: Q, body: T }) => {
  //
}

testFn({
  query,
  // the following `body` emits the lint error
  body,
})
```

**Expected Result**

**Actual Result**

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `X.Y.Z` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `X.Y.Z` |
| `node`                             | `X.Y.Z` |

**Comment by bradzacher** [comment]
It's an unsafe assignment just like it would be if you did `const body = req.body`

Why would we create a special case exemption for destructuring? I'm not sure I understand why it should be treated any differently to a variable assignment.

**Comment by JounQin** [comment]
It's a bit conflict with https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/consistent-destructuring.md

For destrucing, the type comes from a nested object property, but can not use `as` assertion easily.

```ts
const body = req.body as T
const { body } = req // How to?
```

For shorthand, it's just a bit verbose to do `as` assertion IMO.

```ts
testFn({
  body: body as T // I'd prefer shorthand
})
```

**Comment by bradzacher** [comment]
Sorry to clarify - code style is an orthogonal discussion to type safety.

Why should we add an exception for

```ts
const { x } = y as { x: any };
```

When it is just as unsafe as

```ts
const x = (y as { x: any }).x;
```

The important thing here is notifying you about the unsafeness at the source of the issue to ensure its easy to find and fix.

If the argument for an exception is opinionated code style - then I don't think that's a good reason to weaken the rule and the checks it provides.

**Comment by JounQin** [comment]
@bradzacher As I said, the difference is there is no way to use `as` assertion for destructing, and the unsafe warning will emit twice for destructing and assignment.

```ts
const { body } = req // once
testing({
  body, // twice
})
```

**Comment by bradzacher** [comment]
> the difference is there is no way to use as assertion for destructing

You can use `typeof` with an intersection to override the properties you care about.

```ts
declare const x: { a: number, body: any };

{
const {
    a,   // typeof == number
    body // typeof == string
} = x as (typeof x) & { body: string };
}

{
const {
    a,   // typeof == number
    body // typeof == string
}: (typeof x) & { body: string } = x;
}
```

Though I'd **personally** consider eitehr (a) PR to fix the types upstream if they'll allow it. If they don't want it then (b) use something like [patch-package](https://www.npmjs.com/package/patch-package) to fix the types locally.

> and the unsafe warning will emit twice for destructing and assignment

It will emit twice for a variable declaration as well.
The two are literally identical in that regard.

```ts
const body = req.body; // once
test({ body }); // twice
```

> I'm not asking for ignoring it consistently, but provide a new option to ignore it.

You are asking for an option to ignore it consistently.
In a rule about safety I don't think that's at all a good option to add as it inherently introduces unsoundness.

In this **particular** case the same rule emits twice because you're assigning `body` to a variable and then assigning it to a property.

However - if you instead do something else with `body`, then the rule will only emit once, and a different rule (might) emit on that second usage.

```ts
test(body);            // no-unsafe-argument
body + 1;              // restrict-plus-operands
body();                // no-unsafe-call
body.length;           // no-unsafe-member-access
() => {return body};   // no-unsafe-return
`${body}`;             // restrict-template-expressions
if (body) {}           // strict-boolean-expressions
while (body) {}        // strict-boolean-expressions

export { body };       // not covered by anything
export default body;   // not covered by anything
throw body;            // not covered by anything
switch (body) {        // not covered by anything
  case body:           // not covered by anything
}
```

Which is where it's unsound to ignore the destructure - even if you have all our rules turned on, the `any` can 
still leak into your codebase unknowingly. It's a false-negative which can poison your codebase and introduce a bug.

Given that you're the first person  to request this since the rule was released over a year ago - I don't think it's a widely spread issue - so I don't think it is a good option to add.

I don't think adding an option to allow an unsound codestyle in a rule about soundness is a good idea.

For those reasons I am going to reject this proposal.

**Comment by JounQin** [comment]
@bradzacher 

Hah, I've never thought about `const { body } = req as Request & { body: T }`, it should cover the the case already, thanks.
